SE-0036: Requiring Leading Dot Prefixes for Enum Instance Member Implementations

ahoppen opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
Implemented [SE-0036](https://github.com/apple/swift-evolution/blob/master/proposals/0036-enum-dot.md): Requiring Leading Dot Prefixes for Enum Instance Member Implementations

#### Resolved bug number: [SR-1236](https://bugs.swift.org/browse/SR-1236)
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

ahoppen commented:
I also submitted a pull request to swift-package-manager to update its code for the changes: [#260](https://github.com/apple/swift-package-manager/pull/260)

jrose-apple commented:
Interesting. I'm only realizing now that SE-0036 isn't precise enough: does it mean enums have to be accessed as static members (this patch), or that enum cases should never be found via unqualified lookup, even in static methods?

Please add a test with a static method either way.

jrose-apple commented:
Ah, I see that case is actually called out in the "Detail Design" section, and I see your test for it as well now. My bad.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -93,6 +93,9 @@ ERROR(could_not_use_type_member,none,
       "static member %1 cannot be used on instance of type %0",
       (Type, DeclName))
+ERROR(could_not_use_enum_element_on_instance,none,
+      "enum element %0 cannot be referenced as instance members",
```
Singular/plural mismatch here. How about "…as an instance member"?

ahoppen replied:
Sounds good, I’ll change it

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2273,6 +2273,27 @@ diagnoseUnviableLookupResults(MemberLookupResult &result, Type baseObjTy,
+            }
+          }
+          if (triedToAccessEnumElement) {
+            diagnose(loc, diag::could_not_use_enum_element_on_instance,
```
Can we figure out a fix-it here? Just using a dot wouldn't always be correct, but if we could identify some definite circumstances where it was, it'd be great to use it.

ahoppen replied:
I tried adding a fix-it but there wasn’t enough context available to decide whether a leading dot would do (at least as far as I could see). There was also a reason was difficult/impossible as well (I’m just having a look to remember again). Since the fully qualified type was often far more verbose than it really needed to be, so I decided against it.

jrose-apple replied:
Okay. We can always come back and improve it later (possibly by adding more information about the context).

ahoppen replied:
I think, I had an issue with the instantiation of generic parameters, but now that I think about it again, I believe adding a fully qualified type in front should be easily possible. I will add this fix-it together with the changed error message.

ahoppen commented:
I updated the error message and added a fix-it to suggest the full type name. This changed the code in CSDiag.cpp, the error message in DiagnosticsSema.def and of course the test cases in test/Parse/enum.swift. Can you have a look again?

tkremenek commented:
@jrose-apple can you take a look?

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2273,6 +2273,32 @@ diagnoseUnviableLookupResults(MemberLookupResult &result, Type baseObjTy,
+              llvm::raw_string_ostream typeNameStream(enumTypeName);
+              typeNameStream << parentEnum->getNameStr();
+              if (auto genericParams = parentEnum->getGenericParams()) {
+                genericParams->print(typeNameStream);
```
Won't this include constraints and things we don't want? (Please add that to your test either way.)

Actually, in 99% of cases the generic parameters will be assumed anyway, so maybe it'd be better to just leave them out.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2273,6 +2273,32 @@ diagnoseUnviableLookupResults(MemberLookupResult &result, Type baseObjTy,
+            if (auto enumElementDecl = dyn_cast<EnumElementDecl>(decl)) {
+              auto parentEnum = enumElementDecl->getParentEnum();
+              std::string enumTypeName;
+              llvm::raw_string_ostream typeNameStream(enumTypeName);
```
Nitpick: Please use `SmallString` and `llvm::raw_svector_ostream` here.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2273,6 +2273,32 @@ diagnoseUnviableLookupResults(MemberLookupResult &result, Type baseObjTy,
+              auto parentEnum = enumElementDecl->getParentEnum();
+              std::string enumTypeName;
+              llvm::raw_string_ostream typeNameStream(enumTypeName);
+              typeNameStream << parentEnum->getNameStr();
```
and please just use `getName` instead of `getNameStr` here. There's an overload of `<<` for Identifiers.

jrose-apple commented on test/Sema/enum_equatable_hashable.swift:
```diff
@@ -13,8 +13,8 @@ enum Generic<T> {
   func method() -> Int {
-    if A == B { }
-    return A.hashValue
+    if .A == .B { }
```
This isn't really testing the same thing; it can pick *any* `==`. Can you make the method static instead?

jrose-apple commented:
Sorry for the delay!

Aside from my low-level comments, I think always adding the fully-qualified name is the wrong way to go. If a fix-it isn't code that most users would actually accept, we shouldn't offer it. If there's not an easy way to detect whether the fully-qualified name is the right way to go, or whether the single dot would work, we should just leave it out for now.

ahoppen commented:
That was my initial concern as well. I can’t think of a way to determine if a single dot suffices since it depends on the parent of the expression that caused the error and I can’t think of a way to determine it. I just though it would be helpful for the migration assistant to offload the migration work to the IDE instead of the developer.
But since you think otherwise, I removed the fix-it again.

jrose-apple commented:
We can always improve the diagnostic later if we think of something. This is looking good!

@swift-ci Please test

ahoppen commented:
Sorry, missed one declaration in the standard library when resolving merge conflicts. Don’t know why my local tests didn’t catch it. Should be resolved now. Testing locally but am pretty sure it was the only case. Could you try again?

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Probably unrelated?

@swift-ci Please test Linux platform

DougGregor commented:
@swift-ci Please test

ahoppen commented:
I didn’t know there were tests for the swift-pm that could fail because of this. I submitted a pull request for swift-pm to incorporate the changes: https://github.com/apple/swift-package-manager/pull/325

As to the failings on Linux, I’ve got no idea what caused them and don’t have a Linux machine to test it on, but it seems unrelated to my changes to me…

tkremenek commented:
@swift-ci test

jrose-apple commented:
I'm going to ignore the LLDB failures. Thanks, Alex!

bob-wilson commented:
The LLDB failures are breaking a number of the Jenkins bots. Is someone looking into fixing those?

alexwlchan commented:
I’ve been getting CI failure emails that I think are meant for @ahoppen, coming from this PR.

That might be why nobody’s been looking into the failures.

bob-wilson commented:
Todd Fiala just committed a fix for the LLDB failures. I'm following up on fixing the confusion between Alex Hoppen and Alex Chan.

jrose-apple commented:
Ah, then I apologize for assuming they were unrelated! @tfiala, I'll be more careful in the future.

tfiala commented:
No worries, @jrose-apple :-)

tkremenek commented:
I am going to need to revert this.  What I have found from running this change on various code is that there is no migration help.  We provide no fixits for missing dots, `case` statements simply say "invalid pattern" instead of suggesting a dot, etc.  With this kind of breaking change we need better migration support.

tkremenek commented:
For example, for things like this:

~~~swift
 func staticReferencInInstanceMethod() {
    let _ = A // expected-error {{enum element 'A' cannot be referenced as an instance member}}
    let _ = SE0036.A
}
~~~

We should have a Fix-it, not just an error.

Also, for things like this:

~~~swift
switch something {
   case enumCase(let i):
~~~

Right now we get "invalid pattern".  We need a fixit and a better error to help users move forward.

tkremenek commented:
For reference, his was reverted in #2477.

tkremenek commented:
@ahoppen I really appreciate you working on this.  I know this has taken some iteration, but this just needs some more refinement before it can land.  This became apparent when I saw the fallout from landing this change.

jrose-apple commented:
Hm. We *had* a fix-it earlier, and found that it usually didn't match what users would naturally use: it inserted the full name of the enum rather than just a dot.

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -93,6 +93,9 @@ ERROR(could_not_use_type_member,none,
       "static member %1 cannot be used on instance of type %0",
       (Type, DeclName))
+ERROR(could_not_use_enum_element_on_instance,none,
+      "enum element %0 cannot be referenced as an instance member",
```
In fact I think it would be better if the diagnostic just said the leading dot is required

lattner replied:
That only works if in a contextual position where a leading dot will work.

ahoppen commented:
First of all a big thanks to @tfiala for fixing the issues that I caused in LLDB and sorry for all the trouble caused.

This weekend I will go and have a look whether I can come up with an idea on how the determine if a dot suffices in a fix-it that I didn’t think of before. I will report back then.

tfiala commented:
> On May 12, 2016, at 2:27 PM, Alex Hoppen <notifications@github.com> wrote:
> 
> First of all a big thanks to @tfiala <https://github.com/tfiala> for fixing the issues that I caused in LLDB and sorry for all the trouble caused.
> 
No worries, Alex!  It was fine.

I am going to work on improving exposure of build-related failures during an LLDB test run (I.e. when an LLDB test fails to compile a test subject, like a .swift, .cpp or .c program that will be used during testing).  The intent will be to make this show up neatly in the Swift CI test output.  That’ll go a long way to making it more obvious when a Swift change requires modifying some LLDB test subject .swift/.c/.cpp files.
> This weekend I will go and have a look whether I can come up with an idea on how the determine if a dot suffices in a fix-it that I didn’t think of before. I will report back then.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/2224#issuecomment-218891055>

ahoppen commented:
I created an updated pull request with fix its: #2634
